packages/utils/telemetry-utils/src/batchedTelemetryHelper.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type { IDisposable } from "@fluidframework/core-interfaces";
+
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * TODO
+ */
+interface ICustomDataMap {
+	[key: string]: number;
+}
+
+/**
+ * TODO
+ */
+export class BatchedTelemetryHelper implements IDisposable {
+	disposed: boolean = false;
+
+	// Stores value of the custom data passed into the logger.
+	private readonly customDataMap: Map<string, number> = new Map<string, number>();
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	/**
+	 * TODO
+	 */
+	public constructor(
+		private readonly eventBase: ITelemetryGenericEventExt,
+		private readonly logger: ITelemetryLoggerExt,
+		private readonly threshold: number,
+	) {}
+
+	private incrementThresholdCount(): void {
+		this.counter++;
+	}
+
+	private resetThresholdCount(): void {
+		this.counter = 0;
+	}
+
+	private isAboveThreshold(): boolean {
+		return this.counter >= this.threshold;
+	}
+
+	/**
+	 * TODO
+	 */
+	public log(data: ICustomDataMap): void {
+		for (const key of Object.keys(data)) {
+			if (Object.prototype.hasOwnProperty.call(this.customDataMap, key)) {
+				this.customDataMap[key] += data[key];
+			} else {
+				this.customDataMap[key] = data[key];
+			}
+		}
+
+		this.incrementThresholdCount();
+
+		if (this.isAboveThreshold()) {
+			this.sendData();
+		}
+	}
+
+	public sendData(): void {
+		const customData = Object.fromEntries(
+			[...this.customDataMap.entries()].map(([key, value]) => [key, value / this.counter]),
+		);
+
+		// TODO: Add `average` name to the custom data.
+		const telemetryEvent: ITelemetryPerformanceEventExt = {
+			...this.eventBase,
+			...customData,
+		};
+
+		this.logger.sendPerformanceEvent(telemetryEvent);
+		this.resetThresholdCount();
```

Comment:
At this point you also need to clear all the running counters in customMap, otherwise the next iteration is starting from "dirty values", not 0.

Reply:
[5141])https://dev.azure.com/fluidframework/internal/_workitems/edit/5141/)

_"This information should likely be logged in batches and aggregated to avoid too much overhead."_

--

It might be nice to iron out how we want the logger to emit an event. 

Do we want the logger to send telemetry for:
- Every `this.threshold` times over the _entire_ dataset.
- Every `this.threshold` operations that have happened so far and _reset_ the dataset.

Reply:
My take: Every this.threshold operations that have happened so far and reset the dataset.
Otherwise earlier data points get to contribute to multiple rounds of messaging, which gives them an unfairly high weight.